Refactor PlatformInformationProvider

davidmrdavid opened the pull request:
<!-- Start the PR description with some context for the change. -->
Refactors the `PlatformInformationProvider` to utilize `Enum`s more heavily to represent the enviroment. This allows for a pattern-matching-style approach to platform-aware configuration, and requires fewer methods to maintain.

<!-- Make sure to delete the markdown comments and the below sections when squash merging -->
### Issue describing the changes in this PR
N/A

### Pull request checklist

* [x] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation PR is ready to merge and referenced in `pending_docs.md`
* [x] My changes **should not** be added to the release notes for the next release
    * [ ] Otherwise: I've added my notes to `release_notes.md`
* [] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [x] I have added all required tests (Unit tests, E2E tests)

cgillum commented on src/WebJobs.Extensions.DurableTask/DefaultPlatformInformationProvider.cs:
```diff
@@ -49,6 +44,60 @@ public bool InLinuxAppService()
+            return OperatingSystem.Windows;
+        }
+
+        public AppServicePlan GetAppSevicePlan()
```
The name "AppServicePlan" is misleading because it has a very specific meaning in the context of Azure (it refers to a _name_ of a plan, and excludes Consumption and ElasticPremium plan types, which are technically distinct from "App Service plans"). Consider "PlanType" instead for both the name of this enum and the name of this method.

cgillum commented on src/WebJobs.Extensions.DurableTask/DefaultPlatformInformationProvider.cs:
```diff
@@ -49,6 +44,60 @@ public bool InLinuxAppService()
+
+        public AppServicePlan GetAppSevicePlan()
+        {
+            if (this.InLinuxConsumption() | this.InWindowsConsumption())
```
Seems like `||` (boolean operator with shortcircuiting) would be more appropriate than `|` here.

```suggestion
            if (this.InLinuxConsumption() || this.InWindowsConsumption())
```

cgillum commented on src/WebJobs.Extensions.DurableTask/IPlatformInformationService.cs:
```diff
@@ -5,6 +5,64 @@
+    /// <summary>
+    /// Representation of supported Programming Languages.
+    /// </summary>
+    public enum ProgLanguage
```
The name `ProgLanguage` is misleading because the programming language itself is irrelevant. What matters is the stack (.NET, Node.js, Python, PowerShell).

cgillum commented on src/WebJobs.Extensions.DurableTask/DefaultPlatformInformationProvider.cs:
```diff
@@ -49,6 +44,60 @@ public bool InLinuxAppService()
+        public bool IsOutOfProc()
+        {
+            ProgLanguage progLanguage = this.GetProgLanguage();
+            if (progLanguage != ProgLanguage.Csharp)
```
.NET Isolated is also out-of-proc.

davidmrdavid replied:
I think I'll remove this method in that case, as I can just do a more enum comparison instead

cgillum commented on src/WebJobs.Extensions.DurableTask/DefaultPlatformInformationProvider.cs:
```diff
@@ -49,6 +44,60 @@ public bool InLinuxAppService()
+        public ProgLanguage GetProgLanguage()
+        {
+            string workerRuntime = this.nameResolver.Resolve("FUNCTIONS_WORKER_RUNTIME");
+            if (workerRuntime == "python")
```
Use case-insensitive string comparison.

cgillum reviewed (commented):
Some feedback, mostly in regards to naming.

davidmrdavid commented:
In my latest commit, in addition to addressing the feedback so far, I also rename the `PlatformInformationProvider` class itself.

In particular, I remove the `Provider` part and I also removed the `Service` suffix from the interface it implements. I do this because `Provider` has a very specific connotation as a design pattern, which this code does not adhere to. Similarly, the word `Service` doesn't really add any extra meaning to the interface name.

ConnorMcMahon commented on src/WebJobs.Extensions.DurableTask/DefaultPlatformInformation.cs:
```diff
@@ -49,6 +44,44 @@ public bool InLinuxAppService()
+        public WorkerRuntimeType GetWorkerRuntimeType()
+        {
+            string workerRuntime = this.nameResolver.Resolve("FUNCTIONS_WORKER_RUNTIME");
+            if (string.Compare(workerRuntime, "python", ignoreCase: true) == 0)
```
nit: it is more common to see `string.Equals(workerRuntime, "python", StringComparison.OrdinalIgnoreCase)`, but this seems to work too.

ConnorMcMahon replied:
This entire if-else block could probably be represented with [Enum.TryParse](https://docs.microsoft.com/en-us/dotnet/api/system.enum.tryparse?view=net-5.0)

ConnorMcMahon commented on src/WebJobs.Extensions.DurableTask/DefaultPlatformInformation.cs:
```diff
@@ -10,28 +10,23 @@ namespace Microsoft.Azure.WebJobs.Extensions.DurableTask
+        private bool InLinuxConsumption()
         {
             string containerName = this.GetContainerName();
             string azureWebsiteInstanceId = this.nameResolver.Resolve("WEBSITE_INSTANCE_ID");
```
Consider using `Lazy<string>` variables here so that each environment variable is only fetched once, regardless of how many times methods needing this environment variable are called.

ConnorMcMahon commented on src/WebJobs.Extensions.DurableTask/IPlatformInformation.cs:
```diff
@@ -6,45 +6,74 @@
         /// </summary>
-        /// <returns>True if running in Consumption. Otherwise, False.</returns>
-        bool InConsumption();
+        Csharp,
```
rename from `Csharp` to `DotNetInProcess` for the eventual support for the out of process .NET worker.

ConnorMcMahon reviewed (commented):
Mostly looks good, provided some tips for maintainability and performance.

amdeel commented on src/WebJobs.Extensions.DurableTask/DefaultPlatformInformation.cs:
```diff
@@ -49,6 +44,44 @@ public bool InLinuxAppService()
 
+        public OperatingSystem GetOperatingSystem()
+        {
+            if (this.InLinuxConsumption() || this.InLinuxAppService())
```
Is there no single way to determine if something is Linux or Windows?

amdeel commented on src/WebJobs.Extensions.DurableTask/DefaultPlatformInformation.cs:
```diff
@@ -40,7 +35,7 @@ public bool InLinuxConsumption()
-        public bool InLinuxAppService()
+        private bool InLinuxAppService()
         {
             string azureWebsiteInstanceId = this.nameResolver.Resolve("WEBSITE_INSTANCE_ID");
```
You could pull this out in another method to make it clear what azureWebsiteInstanceId and inAppService are used for. I see you use them in both of these Linux methods.

And if this variable determines if it's consumption or not, why do you need to call InLinuxConsumption and InWindowsConsumption in the method IsInConsumption? It seems like you're using this variable to determine that already.

amdeel commented on src/WebJobs.Extensions.DurableTask/IPlatformInformation.cs:
```diff
@@ -6,45 +6,74 @@
         /// </summary>
-        /// <returns>True if running in Linux AppService. Otherwise, False.</returns>
-        bool InLinuxAppService();
+        /// <returns>A ProgLanguage enum.</returns>
```
WorkerRuntimeType. Don't forget to update these.

ConnorMcMahon commented on src/WebJobs.Extensions.DurableTask/DefaultPlatformInformation.cs:
```diff
@@ -0,0 +1,106 @@
+                return workerRuntimeType;
+            }
+
+            throw new Exception("Could not determine worker runtime type.");
```
It may be safer to assume in process if we don't recognize the enum, and just log a warning if we don't recognize the enum. I'm concerned that we may not have 100% uniform enumerations for this environment variable.

If we feel confident in throwing an exception, then a nit would be to make the exception more specific, maybe `InvalidOperationException`.

davidmrdavid replied:
My latest changes include this feedback :)

ConnorMcMahon reviewed (changes requested):
A few small concerns, but otherwise looks good.

davidmrdavid commented on src/WebJobs.Extensions.DurableTask/DefaultPlatformInformation.cs:
```diff
@@ -0,0 +1,106 @@
+        {
+            string workerRuntime = this.ReadEnviromentVariable("FUNCTIONS_WORKER_RUNTIME");
+            WorkerRuntimeType workerRuntimeType;
+            if (Enum.TryParse(workerRuntime, ignoreCase: true, out workerRuntimeType))
```
My newly added test ("WorkerRuntimeTypeFollowsSpec") tests against every worker-runtime value.

I also decided to rename the enum `DotNetInProcess` into simply `dotnet` as I don't think we currently have a clear strategy for how dotnet-in-proc and dotnet-ooproc will be differentiated.

My current preference would be not to couple getting the worker-runtime with determining whether a PL is OOProc or not. Instead, once we have full support for the ooproc dotnet worker, we should figure out how to differentiate between in-proc and ooproc in the dotnet case and to use that to write an `IsOOProc` method in this class.

ConnorMcMahon commented on src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs:
```diff
@@ -422,20 +427,19 @@ void IExtensionConfigProvider.Initialize(ExtensionConfigContext context)
                 linuxLogger = new LinuxAppServiceLogger(writeToConsole: false, containerName, tenant, stampName);
             }
-            else if (inLinuxConsumption)
+            else if (inConsumption)
```
nit: the if clause in this else-if is not necessary. 

In addition, below this, we have a guarantee that linuxLogger is initialized so we don't need the null check.

ConnorMcMahon commented on src/WebJobs.Extensions.DurableTask/IPlatformInformation.cs:
```diff
@@ -6,45 +6,79 @@
         /// </summary>
-        /// <returns>True if running in Consumption. Otherwise, False.</returns>
-        bool InConsumption();
+        Dotnet,
```
nit: prefer casing DotNet.

ConnorMcMahon commented on test/FunctionsV2/OutOfProcTests.cs:
```diff
@@ -371,5 +372,42 @@ public async Task InvokeLocalRpcEndpoint()
+
+            var runtimeType = platformInformation.GetWorkerRuntimeType();
+
+            if (workerRuntime == "haskell")
```
This makes this test fragile. I'm curious if we should just have a HashSet (initialize with StringComparison.OrdinalIgnoreCase) of the approved list of runtimes, and then make this check if the workerRuntime does not exist there.
